### ISSUE TITLE
Link block support does not work when block data is changed in render_block_data

### ISSUE THREAD

**Issue by gaambo** [WordPress/gutenberg#51082]
### Description

The link block support generates the unique class name to use for styling two times: 
1. in `wp_render_elements_support_styles`/`gutenberg_render_elements_support_styles`: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/elements.php#L91 - this runs on **`pre_render_block`**
2. in `wp_render_elements_support` / `gutenberg_render_elements_support`: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/elements.php#L61 - this runs on **`render_block`**

The unique class is generated based on the block data array which is serialized and md5-hashed. The problem is: Between those two hooks the filter hook `render_block_data` runs which can change attributes of the block array. But then the md5-based classname will be different.
That means in 1 a different class name will be added to the element than is used in 2 for generating the CSS rule.

I think the way to go would be to generate the classname only once and store it somewhere to be reused when generating the style rules.

### Step-by-step reproduction instructions

Positive test: 
1. Creat a new post
2. Insert a paragraph and set the link color
3. View post: Color should be applied

Negative test: 
1. Create a plugin or put into your functions.php: 
```php
add_filter('render_block_data', function ($block) {
    $block['test_data'] = true;
    return $block;
});
```
2. Create same post as above
3. View post: Color is not appplied
4. Check: Unique classname added to the paragraph is a different one than used in the inline global styles css

### Environment info

- WordPress 6.2.1; also tested on Gutenberg v15.8.1
- Twenty Twenty Three, Custom Theme

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by gaambo** [comment]
Problem still occurs in current trunk. Steps to reproduce are still valid and stay the same.

**Comment by gaambo** [comment]
Fixed by #59533. Tested & verified via WordPress playground.
